x-pack/plugins/security_solution/public/common/components/markdown_editor/plugins/insight/provider.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { flatten } from 'lodash';
+import type { Filter } from '@kbn/es-query';
+import {
+  FILTERS,
+  isCombinedFilter,
+  isRangeFilter,
+  isPhraseFilter,
+  isPhrasesFilter,
+  isExistsFilter,
+  BooleanRelation,
+} from '@kbn/es-query';
+import type { PhraseFilterValue } from '@kbn/es-query/src/filters/build_filters';
+
+export interface Provider {
+  field: string;
+  excluded: boolean;
+  queryType: string;
+  value: string | number | boolean;
+  valueType?: string;
+}
+
+const buildPrimitiveProvider = (filter: Filter): Provider => {
+  const field = filter.meta?.key ?? '';
+  const excluded = filter.meta?.negate ?? false;
+  const queryType = filter.meta?.type ?? FILTERS.PHRASE;
+  const baseFilter = {
+    field,
+    excluded,
+    queryType,
+  };
+  if (isRangeFilter(filter)) {
+    const gte = filter.query.range[field].gte;
+    const lt = filter.query.range[field].lt;
+    const value = JSON.stringify({ gte, lt });
+    return {
+      ...baseFilter,
+      value,
+      queryType: filter.meta.type ?? FILTERS.RANGE,
+    };
+  } else if (isPhrasesFilter(filter)) {
+    const typeOfParams: PhraseFilterValue = typeof filter.meta?.params[0];
+    return {
+      ...baseFilter,
+      value: JSON.stringify(filter.meta?.params ?? []),
+      valueType: typeOfParams,
+      queryType: filter.meta.type ?? FILTERS.PHRASES,
+    };
+  } else if (isExistsFilter(filter)) {
+    return {
+      ...baseFilter,
+      value: '',
+      queryType: filter.meta.type ?? FILTERS.EXISTS,
+    };
+  } else if (isPhraseFilter(filter)) {
+    const valueType: PhraseFilterValue = typeof filter.meta?.params?.query;
+    return {
+      ...baseFilter,
+      value: filter.meta?.params?.query ?? '',
+      valueType,
+      queryType: filter.meta.type ?? FILTERS.PHRASE,
+    };
+  } else {
+    return {
+      ...baseFilter,
+      value: '',
+      queryType: FILTERS.PHRASE,
+    };
+  }
+};
+
+const nonCombinedToProvider = (filters: Filter[]): Provider[] => {
+  return filters.map((filter) => {
+    return buildPrimitiveProvider(filter);
+  });
+};
+
+export const filtersToInsightProviders = (filters: Filter[]): Provider[][] => {
```

Comment:
This function is a doooozy 😂  Can you just add a comment giving some context here? The name is relatively self evident, but any additional notes could help!

Reply:
Thanks!